src/renderers/dom/client/utils/DOMLazyTree.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright 2015, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule DOMLazyTree
+ */
+
+'use strict';
+
+/**
+ * In IE (8-11) and Edge, appending nodes with no children is dramatically
+ * faster than appending a full subtree, so we essentially queue up the
+ * .appendChild calls here and apply them so each node is added to its parent
+ * before any children are added.
+ *
+ * In other browsers, doing so is slower or neutral compared to the other order
+ * (in Firefox, twice as slow) so we only do this inversion in IE.
+ *
+ * See https://github.com/spicyj/innerhtml-vs-createelement-vs-clonenode.
+ */
+var enableLazy = (
+  typeof document !== 'undefined' &&
+  typeof document.documentMode === 'number'
+  ||
+  typeof navigator !== 'undefined' &&
+  typeof navigator.userAgent === 'string' &&
+  /\bEdge\/\d/.test(navigator.userAgent)
```

Comment:
@patrickkettner @csuwildcat Here in React, I added a user-agent test that catches *all* versions of Edge (and IE) because the DOM implementation in Edge is much slower when reparenting large trees than other browsers. That is, to create the structure

```
<body>
  <article>
    <section>
      <div></div>
    </section>
  </article>
</body>
```

the operations

```
var article = document.createElement('article');
var section = document.createElement('section');
var div = document.createElement('div');
section.appendChild(div);
article.appendChild(section);
body.appendChild(article);
```

are most natural for us, but we are forced to reorder the operations in Edge to be

```
var article = document.createElement('article');
var section = document.createElement('section');
var div = document.createElement('div');
body.appendChild(article);
article.appendChild(section);
section.appendChild(div);
```

in order to get reasonable performance. See the link in the comment above for more details. This has not shipped in a React release yet. I heard that there is a chance this may be fixed soon in Edge – obviously I would prefer to not ship a React with this feature test and would much prefer to put an upper bound on the versions of Edge where we use this workaround. Anything I can do to help get this fixed in Edge soon?